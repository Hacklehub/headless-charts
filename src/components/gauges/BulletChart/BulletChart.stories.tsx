/* eslint-disable @typescript-eslint/no-explicit-any */
import BulletChart from '.';
import { Meta } from '@storybook/react';

const data = 85;

export default {
  title: 'Gauge/BulletChart',
  component: BulletChart,
  tags: ['autodocs'],
} as Meta;

/** Bullet chart is a single metric linear gauge. It shows achievement against 4 different metrics (base, target, threshold, max) */
export const Default = {
  args: {
    data,
    label: 'Sales',
    id: 'bullet-chart-default',
    min: 0,
    base: 50,
    target: 80,
    threshold: 90,
    max: 100,
  },
};

export const Styled = {
  args: {
    ...Default.args,
    id: 'bullet-chart-styled',
    className: 'bg-gray-100 rounded',
    classNameData: 'fill-slate-500',
    classNameTarget: 'fill-blue-500 stroke-blue-500',
    classNameMax: 'fill-red-500',
    classNameThreshold: 'fill-yellow-500',
    classNameBase: 'fill-green-500',
  },
};