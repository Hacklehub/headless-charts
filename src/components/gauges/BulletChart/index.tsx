import React, { useRef } from 'react';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';

import { axisBottom } from 'd3-axis';
import { scaleLinear } from 'd3-scale';
import { select } from 'd3-selection';
import { transition } from 'd3-transition';

export interface BulletChartProps {
  id: string;
  className?: string;
  data: number;
  classNameData?: string;
  label?: string;
  min?: number;
  classNameBase?: string;
  base: number;
  classNameTarget?: string;
  target: number;
  threshold: number;
  classNameThreshold?: string;
  max: number;
  classNameMax?: string;
  margin?: {
    left: number;
    right: number;
    top: number;
    bottom: number;
  };
  axisHeight?: number;
  height?: number;
}

const BulletChart = ({
  id,
  className,
  data = 0,
  classNameData = 'fill-blue-500 stroke-blue-500',
  label = '',
  min = 0,
  classNameBase = 'fill-gray-300 dark:fill-gray-500 dark:stroke-gray-500',
  base,
  classNameTarget = 'fill-black stroke-black dark:fill-white dark:stroke-white',
  target,
  threshold,
  classNameThreshold = `fill-gray-200 stroke-gray-200 dark:fill-gray-600 dark:stroke-gray-600`,
  max,
  classNameMax = `fill-gray-100 stroke-grey-100 dark:fill-gray-700 dark:stroke-gray-700`,
  margin = {
    left: 120,
    top: 10,
    right: 20,
    bottom: 0,
  },

  axisHeight = 20,
  height = 50,
}: BulletChartProps) => {
  const previousData = useRef(0);
  const refreshData = React.useCallback(() => {
    const svg = select(`#${id}`);

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];

    const xFn = scaleLinear()
      .domain([min, max])
      .range([0, width - margin.left - margin.right]);

    const g = svg.append('g');

    g.append('text')
      .text(label)
      .attr(
        'class',
        mergeTailwindClasses(`fill-current stroke-current `, className)
      )
      .attr('text-anchor', 'end')
      .attr('font-size', '0.8em')
      .attr('x', margin.left - 10)
      .attr('y', height - axisHeight - 5);

    const bulletG = g
      .append('g')
      .attr('transform', `translate(${margin.left}, 0)`);

    bulletG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses('fill-current stroke-current ', classNameMax)
      )
      .attr('x', xFn(min))
      .attr('y', 0)
      .attr('width', xFn(max))
      .attr('height', height - axisHeight);

    bulletG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses(`fill-current stroke-current `, classNameThreshold)
      )
      .attr('x', xFn(min))
      .attr('y', 0)
      .attr('width', xFn(threshold))
      .attr('height', height - axisHeight);

    bulletG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses(`fill-current stroke-current `, classNameBase)
      )
      .attr('x', xFn(min))
      .attr('y', 0)
      .attr('width', xFn(base))
      .attr('height', height - axisHeight);

    bulletG
      .append('line')
      .attr('y1', 5)
      .attr('y2', height - axisHeight - 5)
      .attr('x1', xFn(target))
      .attr('x2', xFn(target))
      .attr(
        'class',
        mergeTailwindClasses(`stroke-1 stroke-current `, classNameTarget)
      );

    transition();

    bulletG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses(`fill-current stroke-current `, classNameData)
      )
      .attr('x', xFn(min))
      .attr('y', margin.top)
      .attr('width',  xFn(previousData.current) - xFn(min))
      .attr('height', height - axisHeight - margin.top * 2)
      .transition()
      .duration(1000)
      .attr('width', xFn(data) - xFn(min));

    const xAxisG = g.append('g').attr('class', 'axis--x axis ');

    const xAxis = axisBottom(xFn).ticks(5);

    xAxisG
      .attr('transform', `translate(${margin.left}, ${height - axisHeight})`)
      .call(xAxis);

    previousData.current = data;
  }, [
    axisHeight,
    base,
    className,
    classNameBase,
    classNameData,
    classNameMax,
    classNameTarget,
    classNameThreshold,
    data,
    height,
    id,
    label,
    margin.left,
    margin.right,
    margin.top,
    max,
    min,
    target,
    threshold,
  ]);
  React.useEffect(() => {
    refreshData();
  }, [data, refreshData]);

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(
        defaultChartClassNames,
        `h-12`,
        className
      )}
    />
  );
};

export default BulletChart;
