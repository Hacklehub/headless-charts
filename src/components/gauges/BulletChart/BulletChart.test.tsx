import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import BulletChart from '.';

describe('BulletChart', () => {
  it('renders BulletChart component for snapshot test', () => {
    render(<BulletChart label='Sales'
      id='bullet-chart'
      data={85}
      min={0}
      base={50}
      target={80}
      threshold={90}
      max={100} />);
    const bulletChart = screen.getByTestId('bullet-chart');
    expect(bulletChart).toMatchSnapshot();
  });

  it('renders BulletChart component', () => {
    render(<BulletChart label='Sales'
      id='bullet-chart'
      data={85}
      min={0}
      base={50}
      target={80}
      threshold={90}
      max={100} />);
    const bulletChart = screen.getByTestId('bullet-chart');
    expect(bulletChart).toBeInTheDocument();
  });

})
