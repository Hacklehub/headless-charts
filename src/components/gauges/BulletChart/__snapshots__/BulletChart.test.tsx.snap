// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BulletChart > renders BulletChart component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-12 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="bullet-chart"
  id="bullet-chart"
>
  <g>
    <text
      class=" fill-current stroke-current"
      font-size="0.8em"
      text-anchor="end"
      x="110"
      y="25"
    >
      Sales
    </text>
    <g
      transform="translate(120, 0)"
    >
      <rect
        class=" dark:fill-gray-700 dark:stroke-gray-700 fill-gray-100 stroke-grey-100"
        height="30"
        width="-140"
        x="0"
        y="0"
      />
      <rect
        class=" dark:fill-gray-600 dark:stroke-gray-600 fill-gray-200 stroke-gray-200"
        height="30"
        width="-126"
        x="0"
        y="0"
      />
      <rect
        class=" dark:fill-gray-500 dark:stroke-gray-500 fill-gray-300 stroke-current"
        height="30"
        width="-70"
        x="0"
        y="0"
      />
      <line
        class=" dark:fill-white dark:stroke-white fill-black stroke-black"
        x1="-112"
        x2="-112"
        y1="5"
        y2="25"
      />
      <rect
        class=" fill-blue-500 stroke-blue-500"
        height="10"
        width="0"
        x="0"
        y="10"
      />
    </g>
    <g
      class="axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(120, 30)"
    >
      <path
        class="domain"
        d="M0.5,6V0.5H-139.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-27.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-55.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-83.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-111.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-139.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100
        </text>
      </g>
    </g>
  </g>
</svg>
`;
